lib/dor/text_extraction/aws_provider.rb

```diff
@@ -7,12 +7,21 @@ class AwsProvider
       delegate :client, to: :resource
       attr_reader :aws_client_args
 
-      def initialize(region:, access_key_id:, secret_access_key:)
+      def initialize(region:, access_key_id:, secret_access_key:, role_arn: nil)
         @aws_client_args = {
           region:,
           access_key_id:,
           secret_access_key:
         }
+
+        # role_arn handling deals with e.g.
+        # 'Aws::SQS::Errors::AccessDenied: User: arn:aws:iam::1234567890123:user/uname is not authorized
+        # to perform: sqs:receivemessage on resource: arn:aws:sqs:us-west-2:1234567890123:sul-speech-to-text-done
+        # because no resource-based policy allows the sqs:receivemessage action' when polling for SQS messages
+        return unless role_arn.present?
+
+        # https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SQS/Client.html
+        @aws_client_args[:credentials] = assume_role_credentials(region:, access_key_id:, secret_access_key:, role_arn:)
```

Comment:
I left another note about this below, but one thing I noticed in https://github.com/sul-dlss/speech-to-text/issues/28 is that the assume role operation provides *temporary* session credentials. So I needed to either catch the expired token exception, or not cache the client. I chose to do the latter but maybe I should revisit that.

Reply:
ah, good to know -- i'm not sure how to avoid caching for the watcher.  i basically instantiate a poller by using `Aws::SQS::QueuePoller.new(queue_url, options)`, where `options = { client: aws_provider.sqs }` (see `sqs_watcher.rb`).  and then that poller just runs its `#poll` method on the block that handles "done" messages.

ah, but also, i guess this constructor is caching the `Aws::AssumeRoleCredentials` instance, returned by `assume_role_credentials`, as well as the underlying `Aws::STS::Client` instance.

so i guess there's at least one thing to do here, maybe two:
1. something along the lines of turning `aws_client_args` from an instance variable to a method, one where the `:credentials` field is filled in with a fresh result from `assume_role_credentials` every time `aws_client_args` is called. (this `AwsProvider` class)
2. figure out what error to catch, if need be, in the poller, and renew the token in a `rescue` block, or something. (the `SqsWatcher` class)

Reply:
ok, i turned `aws_client_args` into a private method, and left a note about why the result isn't cached, here's the diff (making `aws_client_args` private broke the old test, but i was never wild about what was there, and i think the updated version is better:

`````diff
diff --git a/lib/dor/text_extraction/aws_provider.rb b/lib/dor/text_extraction/aws_provider.rb
index 4c151f1..9b06129 100644
--- a/lib/dor/text_extraction/aws_provider.rb
+++ b/lib/dor/text_extraction/aws_provider.rb
@@ -5,23 +5,12 @@ module Dor
     # The Application's configured interface to AWS.
     class AwsProvider
       delegate :client, to: :resource
-      attr_reader :aws_client_args
 
       def initialize(region:, access_key_id:, secret_access_key:, role_arn: nil)
-        @aws_client_args = {
-          region:,
-          access_key_id:,
-          secret_access_key:
-        }
-
-        # role_arn handling deals with e.g.
-        # 'Aws::SQS::Errors::AccessDenied: User: arn:aws:iam::1234567890123:user/uname is not authorized
-        # to perform: sqs:receivemessage on resource: arn:aws:sqs:us-west-2:1234567890123:sul-speech-to-text-done
-        # because no resource-based policy allows the sqs:receivemessage action' when polling for SQS messages
-        return unless role_arn.present?
-
-        # https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SQS/Client.html
-        @aws_client_args[:credentials] = assume_role_credentials(region:, access_key_id:, secret_access_key:, role_arn:)
+        @region = region
+        @access_key_id = access_key_id
+        @secret_access_key = secret_access_key
+        @role_arn = role_arn
       end
 
       # @return [::Aws::S3::Bucket]
@@ -46,7 +35,7 @@ module Dor
 
       # @return [::Aws::SQS::Client]
       def sqs
-        @sqs ||= ::Aws::SQS::Client.new(aws_client_args)
+        ::Aws::SQS::Client.new(aws_client_args)
       end
 
       # @return [::Aws::S3::Resource]
@@ -56,6 +45,28 @@ module Dor
 
       private
 
+      attr_reader :region, :access_key_id, :secret_access_key, :role_arn
+
+      # @return [Hash<Symbol,Object>] a Hash with connection and credential info, for use instantiating AWS client objects (e.g. SQS, S3)
+      # @note We don't memoize the client args, because if they contain an Aws::AssumeRoleCredentials instance, that instance's initial token
+      #   can go stale. Making a new credentials hash is not resource intensive, so we just make a fresh one each time.
+      def aws_client_args
+        aws_client_args = {
+          region:, access_key_id:, secret_access_key:
+        }
+
+        # role_arn handling deals with e.g.
+        # 'Aws::SQS::Errors::AccessDenied: User: arn:aws:iam::1234567890123:user/uname is not authorized
+        # to perform: sqs:receivemessage on resource: arn:aws:sqs:us-west-2:1234567890123:sul-speech-to-text-done
+        # because no resource-based policy allows the sqs:receivemessage action' when polling for SQS messages
+        return aws_client_args unless role_arn.present?
+
+        # https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SQS/Client.html
+        aws_client_args[:credentials] = assume_role_credentials(region:, access_key_id:, secret_access_key:, role_arn:)
+
+        aws_client_args
+      end
+
       def assume_role_credentials(region:, access_key_id:, secret_access_key:, role_arn:)
         # https://docs.aws.amazon.com/sdk-for-ruby/v3/developer-guide/setup-config.html#aws-ruby-sdk-credentials-access-token
         Aws::AssumeRoleCredentials.new(
diff --git a/spec/lib/dor/text_extraction/aws_provider_spec.rb b/spec/lib/dor/text_extraction/aws_provider_spec.rb
index 34b9591..fddeb1b 100644
--- a/spec/lib/dor/text_extraction/aws_provider_spec.rb
+++ b/spec/lib/dor/text_extraction/aws_provider_spec.rb
@@ -49,7 +49,7 @@ describe Dor::TextExtraction::AwsProvider do
     end
 
     it 'configures the credentials with a role_arn value' do
-      expect(provider.aws_client_args[:credentials]).to eq assume_role_credentials
+      expect(provider.sqs.config.credentials).to eq assume_role_credentials
       expect(Aws::AssumeRoleCredentials).to have_received(:new).once
     end
   end
`````